playground/backend/internal/db/datastore/datastore_db.go

```diff
@@ -63,19 +69,42 @@ func (d *Datastore) deleteObsoleteSnippets(ctx context.Context, snipKey *datasto
 		Namespace(utils.GetNamespace(ctx)).
 		FilterField("persistenceKey", "=", persistenceKey)
 
-		// At the moment, datastore emulator doesn't allow != filters,
-		// hence this crutches
-		// https://cloud.google.com/datastore/docs/tools/datastore-emulator#known_issues
-		// When it's fixed, post-query filter could be replaced with
-		//
-		// FilterField("__key__", "!=", snipKey)
+	// At the moment, datastore emulator doesn't allow != filters,
+	// hence this crutches
+	// https://cloud.google.com/datastore/docs/tools/datastore-emulator#known_issues
+	// When it's fixed, post-query filter could be replaced with
+	//
+	// FilterField("__key__", "!=", snipKey)
 
 	return d.deleteSnippets(ctx, snippetQuery, snipKey)
 }
 
-// PutSnippet puts the snippet entity to datastore
+// PutSnippet puts the snippet entity to datastore using cloud function proxy
 func (d *Datastore) PutSnippet(ctx context.Context, snipId string, snip *entity.Snippet) error {
 	logger.Debugf("putting snippet %q, persistent key %q...", snipId, snip.Snippet.PersistenceKey)
+
+	var err error
+	if d.externalFunctions != nil {
+		err = d.externalFunctions.PutSnippet(ctx, snipId, snip)
+	}
+	if err != nil || d.externalFunctions == nil {
+		if err != nil {
+			logger.Errorf("Datastore: PutSnippet(): error during the PutSnippet() call to the cloud function, "+
+				"accessing the datastore directly, err: %s\n", err.Error())
+		}
+		if d.externalFunctions == nil {
+			logger.Warnf("Datastore: PutSnippet(): external functions are not set, " +
+				"accessing the datastore directly")
+		}
+		return d.PutSnippetDirect(ctx, snipId, snip)
```

Comment:
Why do we need this fallback? It seems concerning/like it opens us up to the same issues this PR is trying to avoid. Are there ever cases where the fallback is expected/valid?

Reply:
Same question applies below (both this and the private question)

Reply:
The fallback is useful for local development as with it the application will work with datastore emulator regardless of whether functions runtime was also started or not.
In the GKE deployment the fallback will fail as there are no permissions to write to the datastore